doc/api/esm.md

```diff
@@ -313,50 +275,353 @@ If a package has no exports, setting `"exports": false` can be used instead of
+stateless).
+
+The reason that state is an issue is because both the CommonJS and ES module
+versions of the package may get used within an application; for example, the
+user’s application code could `import` the ES module version while a dependency
+`require`s the CommonJS version. If that were to occur, two copies of the
+package would be loaded in memory and therefore two separate states would be
+present. This would likely cause hard-to-troubleshoot bugs.
+
+Aside from writing a stateless package (if JavaScript’s `Math` were a package,
+for example, it would be stateless as all of its methods are static), there are
+three ways to isolate state so that it’s shared between the potentially loaded
+CommonJS and ES module instances of the package:
+
+1. If possible, contain all state within an instantiated object. JavaScript’s
+   `Date`, for example, needs to be instantiated to contain state; if it were a
+   package, it would be used like this:
+
+    ```js
+    import date from 'date';
+    const someDate = new date();
+    // someDate contains state; date does not
+    ```
+
+   The `new` keyword isn’t required; a package’s function can return a new
+   object, or modify a passed-in object, to keep the state external to the
+   package.
+
+1. Isolate the state in one or more CommonJS files that are shared between the
+   CommonJS and ES module versions of the package. For example, if the CommonJS
+   and ES module entry points are `index.cjs` and `index.mjs`, respectively:
+
+    ```js
+    // ./node_modules/pkg/index.cjs
+    const state = require('./state.cjs');
+    module.exports.state = state;
+    ```
+
+    ```js
+    // ./node_modules/pkg/index.mjs
+    export state from './state.cjs';
+    ```
+
+   Even if `pkg` is used via both `require` and `import` in an application (for
+   example, via `import` in application code and via `require` by a dependency)
+   each reference of `pkg` will contain the same state; and modifying that
+   state from either module system will apply to both.
+
+   A package utilizing this pattern would not be usable unmodified in browsers
+   or in other environments that lack support for CommonJS.
+
+1. Write a package where state is stored globally. This is similar to the
+   previous approach, but instead of isolating state within a shared CommonJS
+   file it is attached to the global object, e.g.
+   `globalThis[Symbol.for('pkg@1.2.3:state')]`.
+
+    ```js
+    const stateSymbol = Symbol.for('pkg@1.2.3:state');
+    if (!(stateSymbol in globalThis)) {
+      Object.defineProperty(globalThis, stateSymbol, {value: {}});
+    }
+    const state = globalThis[singletonSymbol];
+    ```
+
+   This has the disadvantage of polluting the global namespace, but it is
+   compatible with non-CommonJS environments such as browsers. It also shares
+   state across multiple loaded numbered versions of the same package, e.g.
+   1.0.0 and 2.0.0, not just CommonJS vs ES module versions.
```

Comment:
I would really prefer if we could avoid encouraging global state as a pattern. Can we remove this whole item entirely please?

Reply:
I’m not sure we should remove it. I think there will be some packages that really need to keep internal state, and should be shared between versions; and also want to be browser-compatible. This is the only solution that satisfies those criteria. The only other “shared state” solution, isolating the state in a CommonJS file, isn’t browser-compatible without build tools.

Reply:
The expectation for browser compatibility is that the `"browser"` field can point to a module that will work in the browser. There is no need for state sharing between the browser and Node.js sources.

Reply:
Yes but if I want to write a package that shares state between the two Node builds, and also has state for a browser build; and I don’t want to write separate logic for the browser version of state management; I think this is the only way.

Reply:
You can write your stateful code as ES modules, transpiling to CommonJS from Babel. The stateful code can then use the ES module sources in the browser and the CommonJS sources in Node. That handles it perfectly fine.

Reply:
> The stateful code can then use the ES module sources in the browser and the CommonJS sources in Node.

The issue is that authors want Node to load the ESM sources in versions of Node that support them. So we need not only ESM sources that work unmodified in both Node and browsers; but _also_ that in Node share state with concurrently loaded CommonJS sources.

I also just don’t think that attaching an object to `globalThis` is so terrible. The docs already call this out as a downside, but it’s one that users are familiar with. They understand the risks.

Reply:
> I also just don’t think that attaching an object to globalThis is so terrible

Man, I _do_ love `window.jQuery`. That has absolutely never caused problems in any shipping webapps.

Reply:
Consider the following app:

```
- src/state/state-management-universal.cjs
- src/entry-node.cjs
- browser/entry-browser.mjs
```

`browser/entry-browser.mjs` can load from `../src/state/...` to load stateful modules from their ES sources, while CJS load will load state relatively and thus get the transpiled state module sources.

The "browser" folder doesn't need transpilation. The `src` folder is compiled into `lib` to contain the same sources as `src` but as CommonJS.

We then set the package.json to provide:

```json
{
  "main": "./lib/entry-node.cjs",
  "exports": {
    ".": {
      "browser": "./browser/entry-browser.mjs",
      "require": "./lib/entry-node.cjs",
      "default": "./src/entry-node.mjs"
    }
  }
}
```

The exact details here are very case specific, but the above is a generic approach nonetheless. I'm not sure why it is up to Node.js documentation to solve and document this case though.

Reply:
I do think we should remove this section entirely, because this is objectively terrible, and as provided above there are alternatives that aren't terrible. I would go so far as blocking on this.